src/compiler/checker.ts

```diff
@@ -6386,6 +6386,8 @@ namespace ts {
             // Only narrow when symbol is variable of type any or an object, union, or type parameter type
             if (node && symbol.flags & SymbolFlags.Variable) {
                 if (isTypeAny(type) || type.flags & (TypeFlags.ObjectType | TypeFlags.Union | TypeFlags.TypeParameter)) {
+                    const declaration = getDeclarationOfKind(symbol, SyntaxKind.VariableDeclaration);

```

Comment:
`getDeclarationOfKind` doesn't walk up any tree, it just iterates over a symbol's declarations. `getDeclarationContainer` walks out of a binding pattern (which is unavoidable, but also not highly nested) but then just does `node.parent.parent.parent` or `node.parent`. I'm not seeing a secondary walk here at all, just static overhead. _Now_ as for weather it's worth doing this just to bail early rather than walking to the top of file... no idea. It probably depends on how deeply nested your code gets. I'll run some perf tests with and without the bailout.